datalad/core/distributed/clone.py

```diff
@@ -636,3 +638,76 @@ def _get_installationpath_from_url(url):
     if path.endswith('.git'):
         path = path[:-4]
     return path
+
+
+def decode_source_spec(spec):
+    """Decode information from a clone source specification
+
+    Parameters
+    ----------
+    spec : str
+      Any supported clone source specification
+
+    Returns
+    -------
+    dict
+      The value of each decoded property is stored under its own key in this
+      dict. By default the following keys are return: 'type', a specification
+      type label {'giturl', 'dataladri', 'ria'}; 'source' the original
+      source specification; 'giturl' a URL for the source that is a suitable
+      source argument for git-clone; 'version' a version-identifer, if present
+      (None else).
+    """
+    # standard property dict composition
+    props = dict(
+        source=spec,
+        version=None,
+    )
+    # common starting point is a RI instance, support for accepting an RI
+    # instance is kept for backward-compatibility reasons
+    source_ri = RI(spec) if not isinstance(spec, RI) else spec
+
+    # scenario switch, each case must set 'giturl' at the very minimum
+    if isinstance(source_ri, DataLadRI):
+        # we have got our DataLadRI as the source, so expand it
+        props['type'] = 'dataladri'
+        props['giturl'] = source_ri.as_git_url()
+    elif isinstance(source_ri, URL) and source_ri.scheme.startswith('ria+'):
+        # parse a RIA URI
+        # check for store related configuration for SSH access
+        if source_ri.scheme == 'ria+ssh':
+            hostname = dlcfg.get(
+                'annex.ria-remote.{}.ssh-host'.format(source_ri.hostname),
+                source_ri.hostname)
+            basepath = dlcfg.get(
+                'annex.ria-remote.{}.base-path'.format(source_ri.hostname),
+                '')
+        else:
+            hostname = source_ri.hostname
+            basepath = ''
+
+        dsid, version = source_ri.fragment.split('@', maxsplit=1) \
+            if '@' in source_ri.fragment else (source_ri.fragment, None)
```

Comment:
my reservations for use of `@` remain: https://github.com/datalad/datalad/issues/2109#issuecomment-573678798  Delaying coming up with generally applicable "schema" for URLs would only bring pains (or even prevent) down the road.

Reply:
But here the `@` is used withn `fragment`, so _after_ `#`.

Reply:
rright, and then for other types of URLs we would need to place it into a fragment explicitly as `http://host/path#@version` even when there is nothing else to go into fragment. 
It might work - we would need anyways to generally urlencode all values for the keys in fragment, so `@` or `;` in e.g. email address or even a path (if present among keys in the fragment) shouldn't end up seen in URL, confusing our parser.
Also having fragment for ria as `#datasetid` makes it probably ria specific -- so we would need per url backend handling or just in general decide that non-keyed element in fragment is the `dataset id`.  So, general "schema" would then be
`#[datasetid][;key=value]*[@version]`
although probably `datasetid` wouldn't be handled by majority of "url backends"?

Reply:
note though, that IIRC `@` is not officially an allowed character within fragment
https://tools.ietf.org/html/rfc3986#section-3.5 :
```
      fragment    = *( pchar / "/" / "?" )
```
but above it says:
```
A fragment identifier component is indicated by the presence of a
number sign ("#") character and terminated by the end of the URI.
```
so it seems that splitting it out from a url is unambiguous and having `@` in it shouldn't confuse any system.

Reply:
Cool, thanks for the detailed assessment!